File: leetcode/743_network_delay_time.cpp
#include <vector>
#include <queue>
#include <limits>
#include <algorithm>

class Solution {
public:
    int networkDelayTime(std::vector<std::vector<int>>& times, int n, int k) {
        // Non-negative edge weights
        std::vector<std::vector<std::pair<int, int>>> graph(n + 1); // 1-indexed
        for (const auto& time : times) {
            int u = time[0];
            int v = time[1];
            int w = time[2];
            graph[u].emplace_back(v, w);
        }
        const int INF = std::numeric_limits<int>::max();
        std::vector<int> dists(n + 1, INF); // store distances for each node
        // if none are inf, that means we can reach all and return sum; otherwise return -1
        dists[k] = 0; // start

        std::priority_queue<
            std::pair<int, int>,
            std::vector<std::pair<int, int>>,
            std::greater<std::pair<int, int>>
        > pq;
        pq.emplace(0, k);

        while (!pq.empty()) {
            auto [d, u] = pq.top();
            pq.pop();

            if (d > dists[u]) { // can only be >=
                continue;
            }

            for (auto [v, w] : graph[u]) {
                if ((d != INF) && (d + w < dists[v])) {
                    dists[v] = d + w;
                    pq.emplace(dists[v], v);
                }
            }
        }

        int res = 0;
        for (size_t i = 1; i < dists.size(); ++i) {
            if (dists[i] == INF) {
                return -1;
            }
            res = std::max(res, dists[i]);
        }
        return res;
    }
};

File: leetcode/743_network_delay_time.py
import heapq
from typing import List

class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        '''
        graph via edges
        edge = u, v, w

        w negative? no
        valid input

        dists = {} stores node: shortest known dist from source to node
        min_heap (dist, node)
        finalized = set()

        while min_heap:
            pop min_heap
            process if not finalized
            iterate through neighbors
                skip neighbor if neighbor is finalized
                if dist_to_curr_node + curr_node_to_neighbor_dist < dists[neighbor]:
                    update dists
                    push to min_heap

        check dists; see if any keys were never updated (cannot reach)
        if so, return -1, otherwise return max(dists.values())
        '''
        graph = {}
        for u, v, dist in times:
            if u in graph:
                graph[u].append((v, dist))
            else:
                graph[u] = [(v, dist)]

        dists = {}
        min_heap = [(0, k)] # (dist, node)
        finalized = set()
        for i in range(1, n + 1):
            dists[i] = float('inf')
        dists[k] = 0

        while min_heap:
            dist, node = heapq.heappop(min_heap)
            if node in finalized:
                continue
            finalized.add(node)
            for neighbor, neighbor_dist in graph.get(node, []):
                if neighbor in finalized:
                    continue
                if dist + neighbor_dist < dists[neighbor]:
                    dists[neighbor] = dist + neighbor_dist
                    heapq.heappush(min_heap, (dist + neighbor_dist, neighbor))

        for node, dist in dists.items():
            if dist == float('inf'):
                return -1

        return max(dists.values())
